### PR TITLE
Update message formating to support markdown

### PR DESCRIPTION
**Pull request by GrayHatter**
Votes to merge (need a majority):
- [x] @grayhatter
- [ ] @Zer0-One 
- [ ] @stal888  
- [ ] @JFreegman 
- [ ] @JX7P 
- [ ] @tux3
- [ ] @iphydf
- [x] @Chuongv
- [ ] @Impyy 

@rugk
@lapfox 
@SkyzohKey

[Review on Reviewable](https://reviewable.io/reviews/tox/tox-client-standard/25)

### CODE REVIEW

File: communication/message_formatting.md [hunk 1]
```diff
@@ -4,12 +4,23 @@ Message Formatting
...
 - **3.5.2** ![](/badge/req.png) A receiving client will make a best-effort to display
-  text exactly as it was sent, and must not transform the text in any way.
+  text exactly as it was intended, and must not materially change the text in any way.
     - The above rule does not apply if malformed UTF-8 text is received. In
       this case, a client may transform the text to any extent necessary to
       make it displayable.
 
-- **3.5.3** ![](/badge/req.png) A sending client must not insert or remove any bytes from
-  the original input.
+- **5.3.3** ![](/badge/rec.png) It's a strong suggestion that messaging clients support
```
Comment (stal888) [inline communication/message_formatting.md:14]: > 5.3.3

pls

File: communication/message_formatting.md [hunk 2]
```diff
@@ -4,12 +4,23 @@ Message Formatting
 - **3.5.1** ![](/badge/req.png) A client must encode emoticons as Unicode emoji.
 
 - **3.5.2** ![](/badge/req.png) A receiving client will make a best-effort to display
-  text exactly as it was sent, and must not transform the text in any way.
+  text exactly as it was intended, and must not materially change the text in any way.
```
Comment (stal888) [inline communication/message_formatting.md:7]: Define "materially changed"?

File: communication/message_formatting.md [hunk 3]
```diff
@@ -4,12 +4,23 @@ Message Formatting
...
+    - ![](/badge/req.png) If the client implments any kind of markdown support, it must 
+      not conflict in any way with the markdown used in this standard.
+    - This Flavor of markdown was chosen because of it's protection of common code syntax.
+    - Clients are free to implement any subset or different markdown or markdown like syntax
+      given it doesn't conflict with this standard.
 
-- **3.5.4** ![](/badge/req.png) Sent messages must be encoded in UTF-8.
+- **3.5.4** ![](/badge/req.png) A sending client must not insert or remove any content/data
+  from any message, of any type. Clients may trim white space from a text messages with the
+  exception of newline and carrage returns, they must not be altered in anyway.
```
Comment (stal888) [inline communication/message_formatting.md]: Spelling: "carriage", "any way"

File: communication/message_formatting.md [hunk 4]
```diff
@@ -4,12 +4,23 @@ Message Formatting
...
+
+- **3.5.5** ![](/badge/req.png) Sent messages must be strictly compatable with UTF-8.
```
Comment (stal888) [inline communication/message_formatting.md]: Spelling: "compatible"

File: communication/message_formatting.md (hunk 2 above)
Comment (GrayHatter) [inline communication/message_formatting.md:7]: appended

File: communication/message_formatting.md [hunk 5]
```diff
@@ -4,12 +4,23 @@ Message Formatting
...
+  links/referneces.
+    - ![](/badge/req.png) If the client implments any kind of markdown support, it must 
+      not conflict in any way with the markdown used in this standard.
+    - This Flavor of markdown was chosen because of it's protection of common code syntax.
+    - Clients are free to implement any subset or different markdown or markdown like syntax
+      given it doesn't conflict with this standard.
 
-- **3.5.4** ![](/badge/req.png) Sent messages must be encoded in UTF-8.
+- **3.5.4** ![](/badge/req.png) A sending client must not insert or remove any content/data
+  from any message, of any type. Clients may trim white space from a text messages with the
```
Comment (stal888) [inline communication/message_formatting.md]: I'd rewrite the point like this:

> Clients may trim extraneous white space (except for carriage returns and new lines)
> from messages. Messages may not be altered in any other way.

File: communication/message_formatting.md [hunk 6]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
       this case, a client may transform the text to any extent necessary to
       make it displayable.
+    - A material change is any change that likely will change the structure, layout,
+      content, meaning, text, or intent of the message.
 
-- **3.5.3** ![](/badge/req.png) A sending client must not insert or remove any bytes from
-  the original input.
+- **5.3.3** ![](/badge/rec.png) It's a strong suggestion that messaging clients support 
+  at the very least a subset of the common markdown syntax. The markdown used in this 
+  standard is a subset of Github Flavored Markdown. With the exclusion of direct
```
Comment (stal888) [inline communication/message_formatting.md:16]: This subset should be defined in a table somewhere in the document.

File: communication/message_formatting.md [hunk 7]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
 
-- **3.5.3** ![](/badge/req.png) A sending client must not insert or remove any bytes from
-  the original input.
+- **5.3.3** ![](/badge/rec.png) It's a strong suggestion that messaging clients support 
+  at the very least a subset of the common markdown syntax. The markdown used in this 
+  standard is a subset of Github Flavored Markdown. With the exclusion of direct 
+  links/referneces.
+    - ![](/badge/req.png) If the client implments any kind of markdown support, it must 
+      not conflict in any way with the markdown used in this standard.
+    - This Flavor of markdown was chosen because of it's protection of common code syntax.
```
Comment (stal888) [inline communication/message_formatting.md:20]: Justification goes in the rationale document

File: communication/message_formatting.md (hunk 6 above)
Comment (GrayHatter) [inline communication/message_formatting.md:16]: I don't want to define a subset here. Subset itself is just a nonexhaustive set of the previous. That way if I only want to implement `**bold** /italics/ _underline_` I can just stop there.

Comment (stal888) [inline communication/message_formatting.md:16]: I think there should be a strict subset that we support, otherwise it potentially conflicts with the alteration clause above.

**Comment by cebe** [comment]
Instead of refering to markdown in general, maybe refer to http://commonmark.org/ ? this at least has spec to refer to instead of some vague definition of markdown.

File: communication/message_formatting.md [hunk 8]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
+    - A material change is any change that likely will change the structure, layout,
+      content, meaning, text, or intent of the message.
 
-- **3.5.3** ![](/badge/req.png) A sending client must not insert or remove any bytes from
-  the original input.
+- **5.3.3** ![](/badge/rec.png) It's a strong suggestion that messaging clients support 
+  at the very least a subset of the common markdown syntax. The markdown used in this 
+  standard is a subset of Github Flavored Markdown. With the exclusion of direct 
+  links/referneces.
+    - ![](/badge/req.png) If the client implments any kind of markdown support, it must
```
Comment (rugk) [inline communication/message_formatting.md:18]: typo: `implments` --> `implements`

File: communication/message_formatting.md [hunk 9]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
       make it displayable.
+    - A material change is any change that likely will change the structure, layout,
+      content, meaning, text, or intent of the message.
 
-- **3.5.3** ![](/badge/req.png) A sending client must not insert or remove any bytes from
-  the original input.
+- **5.3.3** ![](/badge/rec.png) It's a strong suggestion that messaging clients support 
+  at the very least a subset of the common markdown syntax. The markdown used in this 
+  standard is a subset of Github Flavored Markdown. With the exclusion of direct 
+  links/referneces.
```
Comment (rugk) [inline communication/message_formatting.md:17]: typo: `referneces`

File: communication/message_formatting.md [hunk 10]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
+    - ![](/badge/req.png) If the client implments any kind of markdown support, it must 
+      not conflict in any way with the markdown used in this standard.
+    - This Flavor of markdown was chosen because of it's protection of common code syntax.
+    - Clients are free to implement any subset or different markdown or markdown like syntax
+      given it doesn't conflict with this standard.
 
-- **3.5.4** ![](/badge/req.png) Sent messages must be encoded in UTF-8.
+- **3.5.4** ![](/badge/req.png) A sending client must not insert or remove any content/data
+  from any message, of any type. Clients may trim extraneous white space (except for 
+  carriage returns and new lines) from messages. Messages may not be altered in any other way.
```
Comment (rugk) [inline communication/message_formatting.md:26]: > Messages may not be altered in any other way.

So if you `must not insert or remove any content/data` in what additional way can you alter messages?
Additionally these modifications **may** be done as you do not disallow it with `must not`.

File: communication/message_formatting.md [hunk 11]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
+  standard is a subset of Github Flavored Markdown. With the exclusion of direct 
+  links/referneces.
+    - ![](/badge/req.png) If the client implments any kind of markdown support, it must 
+      not conflict in any way with the markdown used in this standard.
+    - This Flavor of markdown was chosen because of it's protection of common code syntax.
+    - Clients are free to implement any subset or different markdown or markdown like syntax
+      given it doesn't conflict with this standard.
 
-- **3.5.4** ![](/badge/req.png) Sent messages must be encoded in UTF-8.
+- **3.5.4** ![](/badge/req.png) A sending client must not insert or remove any content/data
```
Comment (rugk) [inline communication/message_formatting.md:24]: Do you really think such a strict prohibition is necessary?
I mean some clients may improve the message they want to send to other clients.

E.g. think of snippets/shortcuts: If you enter an abbreviation (like `IMO`) and press send it automatically expands this into `in my opinion`. And is this "manipulation" allowed if the user sees it (so the text is changed directly after you typed in `IMO`) and not after you clicked on "Send"?

File: communication/message_formatting.md (hunk 6 above)
Comment (rugk) [inline communication/message_formatting.md:16]: [**@GrayHatter**](https://github.com/GrayHatter) commented on [3. Jan. 2016, 20:57 MEZ](https://github.com/Tox/Tox-Client-Standard/pull/25#discussion_r48696389):

> That way if I only want to implement `**bold** /italics/ _underline_` I can just stop there.

The issue is `_underline_` e.g. conflicts with your spec, because if `_underline_` should be underlined you cannot do this, because in GFM it looks like this: _underline_

**Comment by rugk** [comment]
[**@cebe**](https://github.com/cebe) commented on [3. Jan. 2016, 21:10 MEZ](https://github.com/Tox/Tox-Client-Standard/pull/25#issuecomment-168536869):

> Instead of refering to markdown in general, maybe refer to [http://commonmark.org/](http://commonmark.org/) ? this at least has spec to refer to instead of some vague definition of markdown.

Currently it refers to GitHub Flavoured Markdown.

File: communication/message_formatting.md (hunk 11 above)
Comment (GrayHatter) [inline communication/message_formatting.md:24]: This is a good point, maybe it needs to be rewritten as clients must not transform the content or data following user interaction.

**Comment by cebe** [comment]
> Currently it refers to GitHub Flavoured Markdown.

which has no standard, just a few help documentation examples.

File: communication/message_formatting.md (hunk 2 above)
Comment (Zer0-One) [inline communication/message_formatting.md:7]: > exactly as it was intended

It's not a client's job to read your mind. Re-word this back to "exactly as it was sent" and include exceptions where applicable. Also remove "materially change" and replace it with its definition.

File: communication/message_formatting.md [hunk 12]
```diff
@@ -4,12 +4,25 @@ Message Formatting
...
+  text exactly as it was intended, and must not materially change the text in any way.
     - The above rule does not apply if malformed UTF-8 text is received. In
       this case, a client may transform the text to any extent necessary to
       make it displayable.
+    - A material change is any change that likely will change the structure, layout,
+      content, meaning, text, or intent of the message.
 
-- **3.5.3** ![](/badge/req.png) A sending client must not insert or remove any bytes from
-  the original input.
+- **5.3.3** ![](/badge/rec.png) It's a strong suggestion that messaging clients support
```
Comment (Zer0-One) [inline communication/message_formatting.md:14]: We don't use phrases like "It's a strong suggestion". We use "may" or "shall" and the "rec" and "req" tags to denote a requirement or a recommendation. Re-word this to "A client may support a subset of the" blah blah blah, and optionally reference a table somewhere like stal suggested.

File: communication/message_formatting.md (hunk 2 above)
Comment (GrayHatter) [inline communication/message_formatting.md:7]: exactly as is was sent precludes markdown support

Comment (Zer0-One) [inline communication/message_formatting.md:7]: @GrayHatter which is why i said "include exceptions where applicable"

Comment (GrayHatter) [inline communication/message_formatting.md:7]: also, message splitting should allow the receiving client to recombine the message into one message, instead of 2 as it was sent.

Comment (Zer0-One) [inline communication/message_formatting.md]: include

Comment (Zer0-One) [inline communication/message_formatting.md]: exceptions

Comment (Zer0-One) [inline communication/message_formatting.md]: where

Comment (Zer0-One) [inline communication/message_formatting.md]: applicable

Comment (GrayHatter) [inline communication/message_formatting.md]: need help untwisting your panties there @Zer0-One?

File: communication/message_formatting.md (hunk 9 above)
Comment (netbsduser) [inline communication/message_formatting.md:17]: I found the wording dubious and have adjusted it (with a little semantic change in line with the standard register of the TCS.)

> Messaging clients may support a subset of the common markdown syntax. This shall be defined as a subset of Github Flavored Markdown not including direct links or references.

**Comment by GrayHatter** [comment]
I'm thinking we should take @cebe's suggestion and use commonmark, and then supply our own changes from github that we like.

**Comment by ProMcTagonist** [comment]
@GrayHatter 

Yes, please at least base it on commonmark. We should start with the only well-defined standard for compatibility.

**Comment by stal888** [comment]
Highlighting the TCS team for the record:

@iphydf
@JX7P 
@JFreegman 
@stal888
@Chuongv
@GrayHatter 
@tux3
@Impyy

**Comment by subliun** [comment]
@subliun

**Comment by Zer0-One** [comment]
@GrayHatter Heads-up, I updated your original comment to be a votes checklist for the committee

**Comment by Chuongv** [comment]
Looks good to me :+1:

**Comment by Zer0-One** [comment]
@Chuongv i'll count your vote, but remember that if this PR changes again, you'll have to re-vote (basically, try not to be too eager to cast your vote)

**Comment by DanTheBritish** [comment]
I'm fine with the CommonMark spec since it's more inclusive and explicit than the Github flavour definition (chiming in since I'm tagged).

File: communication/message_formatting.md (hunk 9 above)
Comment (rugk) [inline communication/message_formatting.md:17]: :+1:

**Comment by rugk** [comment]
Nice to see finally CommonMark in this discussion. However [as already said](https://github.com/Tox/Tox-Client-Standard/pull/25#discussion_r48696872) `_underline_` (parsed: _underline_) or `*bold*` (parsed: _bold_) cannot be used and e.g. [the implementation in Ricin](https://github.com/Tox/Tox-Client-Standard/issues/19#issuecomment-168395904) (tracked in https://github.com/RicinApp/Ricin/issues/30) must be changed in this case.
Have a look at [Babelmark](http://johnmacfarlane.net/babelmark2/?text=*bold*+**real+bold**%0A_underline_+__underline__%0A~strikethrough~+~~strikethrough~~%0A%2Fcursive%2F+%2F%2Fcursive%2F%2F) to see this issue.

Part of CommonMark about this: http://spec.commonmark.org/0.23/#emphasis-and-strong-emphasis

Additionally we should specify the version of the spec to use and whether the whole spec must/should be implemented.
Also note that they plan to release a "finalized 1.0 spec" in "early 2016" so we may require this if it is finished.

**Comment by rugk** [comment]
Additionally as fore [links](http://spec.commonmark.org/0.23/#links) we should make clear that they should (or must?) display the whole link in some way when hovering over it, because otherwise people could ["spoof" links](https://github.com/RicinApp/Ricin/issues/30#issuecomment-158765146).

**Comment by stal888** [comment]
Markdown links are extremely complex. I wouldn't recommend making them part of the subset we support.

**Comment by KoKuToru** [comment]
CommonMark and github do support [html](http://spec.commonmark.org/0.23/#raw-html).
I really don't want to have a **webbrowser** for rendering chat messages.

---

How would the receiving client know it's markdown or not ?
If I get `**abc**` from somebody, it can be rendered as  "**abc**" or "**abc**".

Or is markdown required? No plaintext possible.
So if a client wants to send plaintext it must send escaped makdown ?

So messages are never plaintext and always markdown ?

_message splitting with markdown, will be fun.._

**Comment by GrayHatter** [comment]
I'm 100% with @stal888, the risk for abuse for formatting anything but bare links isn't worth it.

@KoKuToru that's one of the trade offs for markdown support. I'm open to suggestions

**Comment by cebe** [comment]
to avoid HTML, a subset of markdown could be agreed on. e.g. only inline elements without HTML.

**Comment by DanTheBritish** [comment]
I don't at all why clients would need to support HTML - what the client uses to render is totally up to it, but the definition here is _what_ should be rendered, not how.

Adding in the definition that HTML parsing is not part of this spec would alleviate that concern.

Link formatting is also a concern, this should only be specific elements of the CommonMark spec, not all of it.

**Comment by rugk** [comment]
IMHO we also should not require clients to implement the whole CommonMark spec. I would define a list of basic things (like _italic_, _bold_ and so on), so that the basic things are always displayed correctly.
And we may suggest clients to implement the CommonMark spec. Note that - if you say only a subset of CommonMark should be implemented - you basically create a new spec while the intention of CommonMark is to unify all these markdown specs.

**Comment by GrayHatter** [comment]
I'm in agreement with rugk on this one. Maybe we should choose the subset
of CommonMark we want, and then required that it formatting be defined by
CommonMark. Then if needed we can handle corners like snake_var_names, or
`inline code segments`.

**Comment by iphydf** [comment]
It's not necessarily reasonable to _require_ a client to display `_italics_` as _italics_. Supporting markdown parsing can be a requirement, but it should be allowed to strip off all formatting and just do its best to display the textual content itself without formatting.

Reasoning: not all terminals and terminal emulators support italics.

I think it is reasonable to require full* markdown (CommonMark) parsing support, but the ability to display the text as intended should be optional. Yes, the client should try, but if it can't (yet) display formatted text, it can simply display the text and still be in compliance.

*) Possibly including well-formed HTML, which is fairly easy to strip off if it's not supported by the display. On the other hand, HTML is not that widely supported in UI text components, so I wouldn't be opposed to excluding it entirely, not the least for security concerns.

**Comment by ProMcTagonist** [comment]
Showing naked markdown is okay, that's the point of markdown.

Showing naked HTML is atrocious.

Who thought it was a good idea to make a markdown spec contain HTML?

**Comment by cebe** [comment]
@ProMcTagonist the [initial markdown description of John Gruber](http://daringfireball.net/projects/markdown/) introduced it as a kind of shortcut notation for HTML so it could contain HTML itself:
http://daringfireball.net/projects/markdown/syntax#html

it is easy to take HTML out of the spec and leave it as a general purpose formatting language. I have written a converter for Markdown to LaTeX for example.

**Comment by rugk** [comment]
@iphydf Good point with the console compatibility. Indeed at least we cannot require a console client to format markdown.

Just pinging @JFreegman, developer of Toxic, about this.

**Comment by cebe** [comment]
> Indeed at least we cannot require a console client to format markdown.

bold, italics and underline are possible even on console. https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_codes

**Comment by rugk** [comment]
Also on the worst console ever (:wink:)? cmd.exe on Windows?

**Comment by cebe** [comment]
afaik no, but who would want to use toxic on a windows cmd.exe?

**Comment by JFreegman** [comment]
@rugk I don't see a good reason to require markdown support, as this isn't a style guideline and not supporting it won't break anything. The point of standardizing markdown is so that clients will know how to parse the inbound text, at which point they can do whatever they want with it. Also, as @iphydf said, some terminal emulators wouldn't be able to support it.

**Comment by rugk** [comment]
> Also, as [@iphydf](https://github.com/iphydf) said, some terminal emulators wouldn't be able to support it.

Can you name some ones except `cmd.exe`?

**Comment by JFreegman** [comment]
@rugk I've tested Gnome terminal, Konsole, uxterm, and urxvt. Only urxvt supports italics by default, with Toxic's version of ncurses. Based on this sample, my assumption is that most don't.

edit: http://superuser.com/questions/204743/terminal-that-supports-the-ansi-italic-escape-code

**Comment by netbsduser** [comment]
This would be a provisional **yes** from me. Although there are a variety of issues with the clauses, this can be resolved later.

File: communication/message_formatting.md (hunk 2 above)
Comment (optimumtact) [inline communication/message_formatting.md:7]: What about - clients may not alter the message in such a way as to change the intent or meaning of the original message, formatting (bold, italics e.t.c.) and structural alterations are allowed where needed for readability.

Comment (optimumtact) [inline communication/message_formatting.md:7]: Overall a lot of these restrictions seem overly strict, I think we should be more permissive about what a client does with messages. Any bad egg clients should be caught quite quickly by public opinion.

**Comment by alexbakker** [comment]
@GrayHatter Do you want to update this PR with the suggested fixes?

**Comment by GrayHatter** [comment]
I will but it won't be for a few weeks.

**Comment by anoadragon453** [comment]
@nurupo brought up a good point on [toxcore/1502](https://github.com/irungentoo/toxcore/pull/1502#issuecomment-173736975): 

```
I wonder if there is any need in a separate message type for markdown. Why the TOX_MESSAGE_TYPE_NORMAL can't be used for it?

Introducing new message type is not a backwards-compatible change, so it shouldn't be done unless there is a concrete reason for adding it. By "not a backwards-compatible change" I mean that client using older toxcore might not display messages of TOX_MESSAGE_TYPE_MARKDOWN type as they are not aware of such type.

So far I see one reason for introducing a separate message type instead of using TOX_MESSAGE_TYPE_NORMAL: users whose message was not meant to be interpreted as markdown but could be interpreted as markdown would be able to chose TOX_MESSAGE_TYPE_NORMAL so that the receiver party would see the message as it is, not interpreted as markdown.
```

So far the main reasoning I've seen for creating and maintaining a new message type is to not interfere with codeblocks/other uses requiring the proposed md syntax. It's fine to display un-parsed markdown to older clients, it's [designed](https://daringfireball.net/projects/markdown/syntax#philosophy) that way.

This doesn't have to be solved with a separate message type (which as mentioned would break all clients without md support), but rather a sequence to prevent parsing within a set of defined characters. 

Similar to [Wikipedia's <nowiki> tag](https://en.wikipedia.org/wiki/Help:Wiki_markup#Nowiki), we could use a non-intrusive character set to escape syntax parsing. My own vote is for double ``'s, but this can be anything really.

TL;DR An entirely new message type isn't necessary for markdown escape; a non-intrusive character sequence is enough and keeps compatibility with existing clients.

Edit: A convincing argument is implementing this functionality in toxcore would force clients to update faster to support it. It would also be better long-term for the core to standardize differentiation between stylized and non-stylized messages, to avoid different clients using different methods to do so.

**Comment by GrayHatter** [comment]
I'm all for that change in toxcore, but the question is, how do we future proof this?

The POINT of markdown is that it's readable in it's raw form. That change in toxcore negates part of the reason for markdown as opposed to any random formatting syntax.

**Comment by anoadragon453** [comment]
Some people were talking about adding syntax parsing for other languages as well (LaTex, ASCIIDOC), however as I see it this is out of the scope of an encrypted messaging protocol.

Before going forward, can we agree that we're only supporting Markdown?

**Comment by GrayHatter** [comment]
No, we can't agree to only support markdown...

We can however agree that this PR will only be about markdown, and markdown will be selected first.

**Comment by anoadragon453** [comment]
Should we not design with scalability in mind then? Should the new message type be designed to accommodate other syntaxs?

Cause if we're allowing some clients to display the message without parsing, this problem becomes a lot simpler with markdown, but not for other formats...

**Comment by tux3** [comment]
So I'm a bit late to the party, but I don't really see the need for a new message type. 

Markdown is made to be readable even without formatting, after all, so at least sending markdown as a normal message would give a good fallback for clients that don't implement markdown. 
And when people use markdown-like tags in chat it's usually for **emphasis**, so it doesn't harm either if we render that as **emphasized** text in normal messages.

**Comment by JFreegman** [comment]
I'm in complete agreement with @tux3. A new message type is unnecessary complexity.

**Comment by nurupo** [comment]
Why are we discussing the message type in here? There is https://github.com/irungentoo/toxcore/pull/1502 for that.

**Comment by SkyzohKey** [comment]
So, any specification about what CommonMark tags should be allowed ?

I propose the following:

| Name | Syntax | Alt syntax | Render |
| --- | :-: | :-: | --: |
| Bolded | `*bolded*` | `**bolded**` | **bolded** |
| Italicized | `/italicized/` | `//italicized//` | _italicized_ |
| Striked | `~striked~` | `~~striked~~` | ~~striked~~ |
| Underlined | `_underlined_` | `__underlined__` | u̲n̲d̲e̲r̲l̲i̲n̲e̲d̲ |
| Quote | `>This is a quote` | _no alt syntax_ | >This is a quote |
| Inline code | &grave;string foo = "bar";&grave; | _no alt syntax_ | `string foo = "bar";` |

I know that this isn't standard CommonMark but it have the merit to be **clear to understand**, for end users.

**28/09/16 EDIT:** Add Inline code tag.

**Comment by GrayHatter** [comment]
I think --Strikethrough-- should remain `~~Strikethrough~~` because that's what it is nearly everywhere else. Other than that.

I'm all for starting with @SkyzohKey's formatting and adding more later.

**Comment by rugk** [comment]
@SkyzohKey
As already said multiple times your suggestion is _not_ CommonMark-compatible.
So we have to decide whether to use CommonMark or a subset of CommonMark or some syntax like you propose.

> I know that this isn't standard CommonMark but it have the merit to be clear to understand, for end users.

This sounds like you want the second. I agree that it is understandable, but it is not even a subset of CommonMark and therefore we cannot suggest clients to use the whole CommonMark implementation as it would break this one.

**Comment by SkyzohKey** [comment]
> therefore we cannot suggest clients to use the whole CommonMark implementation as it would break this one.

@rugk Why would clients need to implement the whole CommonMark when their is only some tags needed ?

**Comment by GrayHatter** [comment]
@rugk no one (I think) is willing to implement all of common mark. Add to that, if we use a subset of common mark, we'd have to define everything in our spec (IIRC that's what @Zer0-One said). 

So if we're only implementing a subsection, and we have to define that subsection, I'm all for starting small and adding later. It won't be hard to add as much as we want later once we get started.

**Comment by GrayHatter** [comment]
And, as much as I like common mark http://commonmark.org/help/ I like the italic markdown from @SkyzohKey better.

**Comment by cebe** [comment]
It's interesting that commonmark describes how `*empth*` will result in `<em>empth</em>` and `**empth**` will result in `<strong>empth</strong>` but it does not define any styling for these elements. `<em>` is mostly rendered italic and `<strong>` is mostly bold but that is not fix in the commonmark spec. These are HTML elements that can be styled differently. Not sure whether commonmark suits as a spec for text formatting outside of the scope of HTML.

**Comment by ProMcTagonist** [comment]
@SkyzohKey's markup is the most intuitive I've ever seen and follows how people use the symbols in plaintext already ( /this/ for emphasis, etc).

I vote for @SkyzohKey's spec over commonmark.<sup> with the exception of that exact quote rendering</sup>

**Comment by stal888** [comment]
I would prefer it if the >s were kept in the quote rendering.

**Comment by SkyzohKey** [comment]
@stal888 The quote rendering is taken from Ricin, clients can render it like they want, but user should see that it's a quote :)

**Comment by SkyzohKey** [comment]
Bump! We need a tag for monospacing text.

**Comment by ProMcTagonist** [comment]
`?

**Comment by SkyzohKey** [comment]
See this: https://github.com/tux3/qTox/issues/2999

**Comment by rugk** [comment]
Closed? Could you give any reason, please?

**Comment by GrayHatter** [comment]
it's gone stale, and I don't have the energy to fight with the Tox Project anymore

You're more than welcome to take this one over
